datahub-web-react/src/app/document/DocumentTreeContext.tsx

```diff
@@ -0,0 +1,358 @@
+ * DocumentTreeContext - Single source of truth for the document tree state.
+ *
+ * This context manages a clean, in-memory tree structure that is completely
+ * decoupled from Apollo cache. All tree operations (move, rename, delete)
+ * update this state immediately, providing instant UI feedback.
+ *
+ * Backend mutations happen as side effects, with rollback on failure.
+ * No more Apollo cache complexity or eventual consistency issues.
+ */
+
+export interface DocumentTreeNode {
+    urn: string;
+    title: string;
+    parentUrn: string | null; // null = root
+    hasChildren: boolean;
+    children?: DocumentTreeNode[]; // Loaded children (undefined = not loaded yet, or merged with server)
+}
+
+interface DocumentTreeContextType {
+    // Tree state
+    nodes: Map<string, DocumentTreeNode>;
+    rootUrns: string[];
+
+    // Tree queries
+    getNode: (urn: string) => DocumentTreeNode | undefined;
+    getRootNodes: () => DocumentTreeNode[];
+    getChildren: (parentUrn: string | null) => DocumentTreeNode[];
+
+    // Tree mutations (immediate state updates)
+    updateNodeTitle: (urn: string, newTitle: string) => void;
+    moveNode: (urn: string, newParentUrn: string | null) => void;
+    deleteNode: (urn: string) => void;
+    addNode: (node: DocumentTreeNode) => void;
+    setNodeChildren: (parentUrn: string | null, children: DocumentTreeNode[]) => void;
+
+    // Batch initialization (for loading root documents or initial data)
+    initializeTree: (rootNodes: DocumentTreeNode[]) => void;
+}
+
+export const DocumentTreeContext = React.createContext<DocumentTreeContextType | undefined>(undefined);
+
+export const useDocumentTree = () => {
+    const context = useContext(DocumentTreeContext);
+    if (!context) {
+        throw new Error('useDocumentTree must be used within DocumentTreeProvider');
+    }
+    return context;
+};
+
+export const DocumentTreeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+    const [nodes, setNodes] = useState<Map<string, DocumentTreeNode>>(new Map());
+    const [rootUrns, setRootUrns] = useState<string[]>([]);
+
+    // Query: Get a single node
+    const getNode = useCallback(
+        (urn: string) => {
+            return nodes.get(urn);
+        },
+        [nodes],
+    );
+
+    // Query: Get all root nodes
+    const getRootNodes = useCallback(() => {
+        return rootUrns.map((urn) => nodes.get(urn)).filter(Boolean) as DocumentTreeNode[];
+    }, [nodes, rootUrns]);
+
+    // Query: Get children of a parent (or root if parentUrn is null)
+    const getChildren = useCallback(
+        (parentUrn: string | null) => {
+            if (parentUrn === null) {
+                return getRootNodes();
+            }
+            const parent = nodes.get(parentUrn);
+            return parent?.children || [];
+        },
+        [nodes, getRootNodes],
+    );
+
+    // Mutation: Update node title
+    const updateNodeTitle = useCallback((urn: string, newTitle: string) => {
```

Comment:
okay gotcha so the pattern i assume will be that this context is simply managing state, then mutations happen at the lower level at the same place where we are actually updating

i could see a world where we combine them in these contexts to consolidate everything about mutations/reading for a given thing (ie. Document Tree here) in one place and the clients of these contexts don't need to worry about implementation under the hood.

like i see you have `useUpdateDocumentTitleMutation` - we could bring that into the context here

that being said the way it's abstracted into a reusable hook is still nice and reusable so this is still a solid pattern